sycl/doc/extensions/proposed/sycl_ext_oneapi_address_space_cast.asciidoc

```diff
@@ -0,0 +1,192 @@
+whether a given raw pointer can be cast to a specific address space; and 2)
+performs the casting operation. In cases where the developer is attempting to
+assert that a raw pointer points to an object in a specific address space, the
+checks from 1) are not required and may have undesirable performance impact.
+
+This extension separates `sycl::address_space_cast` into two functions:
+
+- `static_address_space_cast`, which casts with no checks.
+- `dynamic_address_space_cast`, which casts with checks.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_ADDRESS_SPACE_CAST` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Address space cast functions
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+static_address_space_cast(ElementType* ptr);
+
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+dynamic_address_space_cast(ElementType* ptr);
+
+} // namespace sycl::ext::oneapi::experimental
+----
+
+[source,c++]
+----
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+static_address_space_cast(ElementType* ptr);
+----
+_Preconditions_: `ptr` points to an object allocated in the address space
+designated by `Space`.
+
+_Returns_: A `multi_ptr` with the specified address space and decoration that
+points to the same object as `ptr`.
+
+[source,c++]
+----
+template <access::address_space Space, access::decorated DecorateAddress,
+          typename ElementType>
+multi_ptr<ElementType, Space, DecorateAddress>
+dynamic_address_space_cast(ElementType* ptr);
+----
+_Preconditions_: `ptr` is accessible to the calling work-item.
+
+_Returns_: A `multi_ptr` with the specified address space and decoration that
+points to the same object as `ptr` if `ptr` points to an object allocated in
+the address space designated by `Space`, and `nullptr` otherwise.
+
+_Diagnostics_: Implementations using address space deduction must issue a
+diagnostic if `ptr` does not point to an object allocated in the address space
+designated by `Space`.
```

Comment:
Just to be clear -- are you suggesting this should apply to `static_address_space_cast` _instead_ of `dynamic_address_space_cast`?

This could be another example of the SYCL 2020 version of the function conflating two things... It definitely seems reasonable that the "static" version would fail (because the implementation would know the assertion was invalid), while the dynamic version would succeed (because the implementation would know to return `nullptr`).

Reply:
> are you suggesting this should apply to static_address_space_cast instead of dynamic_address_space_cast?

yes, but it is currently under `dynamic_address_space_cast`.

>  It definitely seems reasonable that the "static" version would fail (because the implementation would know the assertion was invalid), while the dynamic version would succeed (because the implementation would know to return nullptr).

Exactly for this reason (and also why I like this ext :). When it is the `static` variant, you are basically asserting so make sense to issue a diagnostic. The `dynamic` variant can be used to test and branch, so emitting a diagnostic isn't right as replacing with nullptr or forwarding the value is a perfectly valid behavior.

Reply:
I agree with @Naghasan that we do *not* want `dynamic_address_space_cast` to issue a diagnostic for implementations that have address space inference.  Consider code like:

```
if (!dynamic_address_space_cast<access::address_space::global_space>(ptr)) {
  /* assume the ptr is not global space */
}
```

The application is only checking the address space here, so why would we issue a diagnostic?  The implementation should just return `nullptr` if it can infer that the address space is not `global_space`.

I think we do *not* want to mandate a diagnostic in the `static_address_space_cast` case either.  Shouldn't we just leave it as QoI, and let the implementation issue a diagnostic for unsafe things if it is able to prove that they are unsafe?

Reply:
> I think we do not want to mandate a diagnostic in the static_address_space_cast case either. Shouldn't we just leave it as QoI, and let the implementation issue a diagnostic for unsafe things if it is able to prove that they are unsafe?

I'm not sure if this should be QoL in this case, because it's so closely tied to the address space inference rules.

[Section 5.9.4](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:inferredAddressSpace) says "It is illegal to assign a pointer value addressing one address space to a pointer variable addressing a different address space." (presumably requiring some kind of diagnostic).  If an assignment is guaranteed to trigger a diagnostic, it would be weird if `static_address_space_cast` behaved differently.

Note that the intent of the wording I proposed here is to only require a diagnostic in the inferred address space mode, which already requires the implementation to be tracking the address space of all pointers (i.e. there aren't really any generic pointers).  I don't think it would be any additional implementation burden to issue a diagnostic in that case.  Implementations supporting generic as the default address space would not be required to issue a diagnostic (but could if they wanted to, because the precondition makes invalid casts UB).

Reply:
> in the inferred address space mode [...] I don't think it would be any additional implementation burden to issue a diagnostic in that case.

I agree, It shouldn't be additional burden as the approach is built is do this anyway.

Reply:
I defer to @Naghasan here, but it seems like an interesting question to me.  Let's consider a case where the implementation uses "inferred address space" mode and it is not able to infer the address space of a pointer.  The user calls `static_address_space_cast`.  Which of these two behaviors should we have:

* Implementation diagnoses an error because it cannot prove the cast is correct, or
* Implementation treats the cast as an assertion by the user and assumes it is correct.

In the "generic as default" mode we assume the cast is correct, and the behavior is undefined if the cast is wrong.  Why should it be different for "inferred address space" mode?

Reply:
@Naghasan: Are you okay with things being merged with the non-normative note about the diagnostic?